Enable NSCF in hp and KCW

gcistaro opened the pull request:
The nscf gives reasonable agreement with all the material tested, with and without MPI.

gcistaro committed:
Error in Cholesky factorization

Commit:
creating new kset and ground state for nscf

toxa81 committed:
fixes

gcistaro committed:
Moving nscf to do_nscf, getting H0 for each q

gcistaro committed:
Moving nscf, calculating H0 for each iq

gcistaro committed:
Fixing error in kcw_run_nscf

gcistaro committed:
enabling number of steps

gcistaro committed:
Removing error files from branch

gcistaro committed:
removing unused variable in kcw_run_nscf

gcistaro committed:
enabling nscf in HP

toxa81 committed:
try 200 steps

gcistaro committed:
temporarely disabling SIRIUS nscf in HP for checking ci-tests

gcistaro committed:
temporarely disabling SIRIUS nscf in HP for checking ci-tests --- correct typo

gcistaro committed:
Fixing error from setup_sirius when nscf is not enabled

toxa81 committed:
try without new handlers

toxa81 committed:
cleanup

gcistaro committed:
Fixing reading of v>1 upf file, when n is not specified in it

gcistaro committed:
remove blank line added by mistake

gcistaro committed:
Merge branch 'fix/read-upf' into Transfer_kpoints_to_SIRIUS_v2

gcistaro committed:
adding header when nscf is done with SIRIUS

gcistaro committed:
Fixing issue with MPI and the et array. Now et is filled with the right eigenvalues.

gcistaro committed:
Fixing eigenvalues storage in qe. et needs all the eigenvalues for rank=0, only local eigenvalues for rank!=0;
Simplifying get_wave_functions_from_sirius: now the array k_index_map is not needed and we added an if to understand if the k point in the loop is local

gcistaro committed:
enabling nscf in HP

gcistaro committed:
restoring previous version of get_wave_functions_from_sirius

gcistaro committed:
restarting CI

gcistaro committed:
Fixing inconsistency in get_wave_functions (it was now defined without arguments

gcistaro committed:
adding parallel testing for KCW

gcistaro committed:
ask 4 nodes for H2O parallel test

gcistaro committed:
Adding kcw.yml file to compare with references in CI

gcistaro committed:
try to fix parallel KCW tests

Commit:
commenting line to hard set use_sirius_scf

gcistaro committed:
Merge branch 'ristretto' into SIRIUS-enabled_nscf_in_LR

toxa81 committed:
enforce real occupcation matrix in sirius

gcistaro committed:
Merging ristretto

toxa81 committed:
Merge branch 'fix/enable_real_occupation_matrix' into SIRIUS-enabled_nscf_in_LR

gcistaro committed:
Fixing bug when more than a perturbation is perturbed, by appropriately setting up sirius in this case

toxa81 committed:
Merge branch 'SIRIUS-enabled_nscf_in_LR' of github.com:electronic-structure/q-e-sirius into SIRIUS-enabled_nscf_in_LR

toxa81 commented:
@gcistaro why there are so many failing ci/cd jobs?

toxa81 commented on PW/src/mod_sirius.f90:
```diff
@@ -78,6 +78,7 @@ MODULE mod_sirius
   !
   TYPE(sirius_kpoint_set_handler) :: ks_handler
   !! SIRIUS k-point set handler
+  TYPE(C_PTR) :: error_code
```
Is this used somewhere? If not -> remove

toxa81 commented on PW/src/mod_sirius.f90:
```diff
@@ -1494,39 +1498,51 @@ SUBROUTINE invert_mtrx(vlat, vlat_inv)
   !
   !-------------------------------------------------------------------------
-  SUBROUTINE get_band_energies_from_sirius()
+  SUBROUTINE get_band_energies_from_sirius(ks_handler_)
```
_ underscore in the parameter name is not nice. To avoid potential name conflict with ks_hangler global variable, here a shorter `ks_h` name can be used. The same applies to `gs_handler_` -> `gs_h`

toxa81 commented on HP/src/hp_main.f90:
```diff
@@ -141,8 +133,18 @@ PROGRAM hp_main
         ! If necessary the bands are recalculated
         !
+#if defined(__SIRIUS)
+        IF ( .not. setup_pw ) THEN
```
What is the logic of `.not. setup_pw` here?

gcistaro replied:
To do nscf I am clearing up and resetting sirius at each iteration (mainly for the kset_handler, k+q is changing and I need to know it for nscf). This is set in PW/src/non_scf.f90 if I need to do the nscf, if not this part is skipped and I need to force the setup_sirius here with this if. I built up this logic so that we can use what is inside non_scf.f90 also using pw with calculation='nscf'. 

If I don't put this line there, we will face the breaking for the second perturbed atom. 
I don't know if it is worth to factorize a part of setup_sirius to do this, in principle what I need to do is just clean up kset_handler and reinitialize it, not the whole sirius interface.

toxa81 commented on LR_Modules/setup_nscf.f90:
```diff
@@ -151,6 +153,20 @@ SUBROUTINE setup_nscf ( newgrid, xq, elph_mat )
           npk, k1, k2, k3, nk1, nk2, nk3, nkstot, xk, kunit)
   END IF
   !
+#if defined(__SIRIUS)
```
This is a copy-paste of the code from setup.f90. Better to move it to a standalone function.

toxa81 commented on LR_Modules/setup_nscf.f90:
```diff
@@ -186,6 +202,7 @@ SUBROUTINE setup_nscf ( newgrid, xq, elph_mat )
   !
   qnorm = sqrt(xq(1)**2 + xq(2)**2 + xq(3)**2) * tpiba
   !
+
```
not needed

gcistaro commented:
> @gcistaro why there are so many failing ci/cd jobs?

@toxa81 It seems parallel jobs get NaN in the linear system but it was not happening before. Did some of the logic change? ( wavefunctions - eigenvalues... )

gcistaro committed:
Cleaning up the code:
- renaming variables from ks_handler_ to ks_h and from gs_handler_ to gs_h
- Removing unwanted empty lines
- factorizing functionalities to avoid code repetions (to setup kpoints and wkpoints in mod_sirius)
- Removing unwanted variables

toxa81 committed:
pass T

toxa81 committed:
Merge branch 'SIRIUS-enabled_nscf_in_LR' of github.com:electronic-structure/q-e-sirius into SIRIUS-enabled_nscf_in_LR

toxa81 committed:
flip sign of the non-local occupation matrix if needed

gcistaro committed:
Setting evc to zero in get_wave_functions_from_sirius to avoid NaN in the elements out of bound.

toxa81 commented:
NiO-ldapuv-ortho test doesn't pass for native q-e 7.3.1, something has changed for the hubbard stress component

toxa81 committed:
touch file to trigger ci/cd run

toxa81 committed:
revert sign in atomic_wfc_mod.f90

toxa81 committed:
add ZrN test

toxa81 committed:
temporary disable reframe; run nscf with qe

gsavva committed:
add FeO test (magnetic metal)

toxa81 committed:
Merge branch 'test_ci_cd' into SIRIUS-enabled_nscf_in_LR

toxa81 committed:
run nscf by qe

gcistaro committed:
Fixing mistake in kcw for calculating the screening coefficients

gcistaro committed:
Adding scripts to run_parallel jobs in KCW

gcistaro committed:
adding one more line on hp_diff output

gcistaro committed:
change folder on Si KCW test to Si_koopmans_parallel

gcistaro committed:
change folder of koopmans parallel in .ci

gcistaro committed:
removing duplicated line in run_parallel.x for Silicon

gsavva committed:
refine tests

- make FeO parallel
- Decrease verbosity of hp to 2.
- In LiCoO2 parallel do the "cat" and check once, by SLURM_PROCID

gsavva committed:
add test CaFeO2 (2-step scf, U+V!=0, magnetic insulator)

gsavva committed:
missed the PP due to .gitignore entry :-\

gcistaro committed:
Adding comparison in run_parallel.x, executing wannier90.x PP only in serial

gcistaro committed:
Merge branch 'SIRIUS-enabled_nscf_in_LR' of github.com:electronic-structure/q-e-sirius into SIRIUS-enabled_nscf_in_LR

gsavva committed:
fixing directories in parallel tests

gcistaro committed:
Adding HP/PW in name of tests; try to fix parallel KCW

gcistaro committed:
Merge branch 'SIRIUS-enabled_nscf_in_LR' of github.com:electronic-structure/q-e-sirius into SIRIUS-enabled_nscf_in_LR

gcistaro committed:
Adding script for kcw where I sort the values before comparing

gcistaro committed:
Adding round to show only a few digits in the comparison of the yml files

gsavva committed:
reduce resources for CaFeO2 test (6 to 4 nodes)

gsavva committed:
add NiO test (AFM insulator), US-PP, pbesol

toxa81 commented:
cscs-ci run

gsavva commented:
@toxa81 any ideas why we get some random (?) failures in the ci tests?

`(HP) Ni parallel` passes [here](https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/6529227236527536/1513296724725430/-/jobs/7428650064) , but [here](https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/6529227236527536/1513296724725430/-/jobs/7434892435) the 1st scf iteration in native QE fails without any obvious error ... Nothing changed in-between the two revisions, I just added once more test!

In another case, [CaFeO2](https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/6529227236527536/1513296724725430/-/jobs/7428650042), native QE did not converge many bands and exited.

> NiO-ldapuv-ortho test doesn't pass for native q-e 7.3.1, something has changed for the hubbard stress component

I tried this locally and I see no issues though... shall we re-enable the ReFrame tests then ?

gsavva commented:
cscs-ci run

toxa81 commented:
I don't see any obvious reason why some jobs fail. This is very annoying. I suspect some race condition in openblas or other part of code. Also microarchitecture: we build for x86_64 but do we know 100% that spack is doing it properly? 
Re: ReFrame. Any time, but this will cost +1,5H of build time for any new change in PR.

gsavva commented:
cscs-ci run

gsavva commented:
cscs-ci run

gsavva commented:
cscs-ci run

gsavva committed:
try with 1 OpenMP thread in an effort to find why native QE scf fails randomly

gsavva commented:
cscs-ci run

gsavva committed:
reduce npool for failing CI tests

gsavva commented:
cscs-ci run

gsavva committed:
revert to 12 OpenMP threads, doesn't seem to affect ci

gsavva committed:
modify configuration of test (HP) Ni parallel: 3 nodes, 3 MPI ranks, 3 kpools

gcistaro committed:
Trying to fix h2o parallel with 3 nodes, distributed over plane waves

gsavva commented:
cscs-ci run

gcistaro committed:
Uncommenting line in h2o scf

gcistaro committed:
regenerating kcw.ref.yml using vanilla qe to compare against

gsavva committed:
Ni parallel: move commands inside script

gcistaro committed:
Add LiF example for KCW

gcistaro committed:
Delete plottings in wannier input

gsavva commented:
cscs-ci run

gsavva commented:
cscs-ci run

gsavva commented:
cscs-ci run

gsavva commented:
cscs-ci run

gsavva committed:
add one more test for HP: non-magnetic metal, U,V != 0, US+PAW PP

gsavva commented:
cscs-ci run

gsavva committed:
use native QE 7.2 Hubbard values as reference

toxa81 reviewed (approved):
All tests are green now. I think it's time to merge it

gsavva commented:
indeed, it's time to merge it (even though it's not clear what causes tests to fail on the QE-side).
Let's add the comparison of the Hubbard V values in a separate branch/PR
